### PR TITLE
BUG: Help python csv engine read binary buffers

### LINKED ISSUE #23779: read_csv c engine accepts binary mode data and python engine rejects it
#### Code Sample

```python
import pandas as pd

if __name__ == "__main__":
    with open('test.csv', 'w') as f:
        f.write('1,2,3\n4,5,6')
    with open('test.csv', 'rt') as f:
        pd.read_csv(f, header=None)
    with open('test.csv', 'rb') as f:
        pd.read_csv(f, header=None)
    with open('test.csv', 'rt') as f:
        pd.read_csv(f, header=None, engine='python')
    with open('test.csv', 'rb') as f:
        pd.read_csv(f, header=None, engine='python')
```
#### Problem description

The second read_csv call (using the C engine and a file opened in binary mode) will correctly read the csv. The fourth read_csv call (using the Python engine and a file opened in binary mode) will throw an exception stating it needs to be in text mode:

```
pandas.errors.ParserError: iterator should return strings, not bytes (did you open the file in text mode?)
```

Perhaps this is intended behavior, but I found this difference in behavior between the engines surprising, as well as that binary mode was accepted at all.

#### Expected Output

Either the C engine rejecting binary mode files or the Python engine accepting them.

#### Output of ``pd.show_versions()``

<details>
INSTALLED VERSIONS
------------------
commit: None
python: 3.6.6.final.0
python-bits: 64
OS: Linux
OS-release: 4.15.0-39-generic
machine: x86_64
processor: x86_64
byteorder: little
LC_ALL: None
LANG: en_US.UTF-8
LOCALE: en_US.UTF-8

pandas: 0.23.4
pytest: None
pip: 10.0.1
setuptools: 39.1.0
Cython: None
numpy: 1.15.4
scipy: None
pyarrow: None
xarray: None
IPython: None
sphinx: None
patsy: None
dateutil: 2.7.5
pytz: 2018.7
blosc: None
bottleneck: None
tables: None
numexpr: None
feather: None
matplotlib: None
openpyxl: None
xlrd: None
xlwt: None
xlsxwriter: None
lxml: None
bs4: None
html5lib: None
sqlalchemy: None
pymysql: None
psycopg2: None
jinja2: None
s3fs: None
fastparquet: None
pandas_gbq: None
pandas_datareader: None
</details>

### PR DESCRIPTION
**Pull request by fiendish**
The file buffer given to read_csv could have been opened in
binary mode, but the python csv reader errors on binary buffers.

- [x] closes #23779
- [x] tests added / passed
- [x] passes `black pandas`
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

### CODE REVIEW

**Comment by fiendish** [comment]
What does Linux py37_np_dev know that I don't?

File: doc/source/whatsnew/v0.25.1.rst [hunk 1]
```diff
@@ -105,7 +105,7 @@ I/O
 ^^^
 
 - Avoid calling ``S3File.s3`` when reading parquet, as this was removed in s3fs version 0.3.0 (:issue:`27756`)
--
+- read_csv now accepts binary mode file buffers when using the Python csv engine (:issue:`23779`)
```
Comment (jreback) [inline doc/source/whatsnew/v0.25.1.rst]: use ``:meth:`read_csv` ``; move this note to 1.0

File: pandas/tests/io/parser/test_python_parser_only.py [hunk 2]
```diff
@@ -296,3 +296,10 @@ def test_malformed_skipfooter(python_parser_only):
     msg = "Expected 3 fields in line 4, saw 5"
     with pytest.raises(ParserError, match=msg):
         parser.read_csv(StringIO(data), header=1, comment="#", skipfooter=1)
+
+
+def test_binary_buffer(python_parser_only, csv1):
```
Comment (jreback) [inline pandas/tests/io/parser/test_python_parser_only.py]: run this for both csv engines

File: pandas/tests/io/parser/test_python_parser_only.py [hunk 3]
```diff
@@ -296,3 +296,10 @@ def test_malformed_skipfooter(python_parser_only):
...
+    # see gh-23779
+    parser = python_parser_only
+    with open(csv1, "rb") as f:
+        parser.read_csv(f)
```
Comment (jreback) [inline pandas/tests/io/parser/test_python_parser_only.py]: assert the result contents

Comment (jreback) [reply pandas/tests/io/parser/test_python_parser_only.py]: move this test to parser/test_common.py see if you can co-locate with other buffer type tests.

**Review by jreback — CHANGES_REQUESTED** [review]

**Comment by fiendish** [comment]
Made the requested changes. Do you want a rebase to squash?

File: pandas/io/common.py [hunk 4]
```diff
@@ -342,9 +342,11 @@ def _get_handle(
     try:
         from s3fs import S3File
 
-        need_text_wrapping = (BytesIO, S3File)
+        need_text_wrapping = (BufferedIOBase, BytesIO, S3File)
```
Comment (TomAugspurger) [inline pandas/io/common.py]: BytesIO is a subclass of BufferedIOBytes. Can BytesIO be removed?

**Review by TomAugspurger — COMMENTED** [review]
No need to squash.

File: pandas/io/common.py (hunk 4 above)
Comment (fiendish) [reply pandas/io/common.py]: 👍

**Comment by fiendish** [comment]
Is it normal for random unrelated numpy errors to fail builds?

**Comment by TomAugspurger** [comment]
Yes, we're debugging it still. Restarted azure.

File: pandas/io/common.py [hunk 5]
```diff
@@ -420,8 +422,10 @@ def _get_handle(
     if is_text and (compression or isinstance(f, need_text_wrapping)):
         from io import TextIOWrapper
 
-        f = TextIOWrapper(f, encoding=encoding, newline="")
-        handles.append(f)
+        g = TextIOWrapper(f, encoding=encoding, newline="")
+        if not isinstance(f, no_close):
```
Comment (jreback) [inline pandas/io/common.py]: this is hard to follow, can you just use BufferedIOBase here rather than no_close?

File: pandas/tests/io/parser/test_common.py [hunk 6]
```diff
@@ -2020,9 +2020,32 @@ def test_file_handles_with_open(all_parsers, csv1):
...
+        ("unicode_series.csv", "latin-1"),
+        ("sauron.SHIFT_JIS.csv", "shiftjis"),
+    ],
+)
+def test_binary_mode_file_buffers(all_parsers, csv_dir_path, fname, encoding):
+    # gh-23779: Python csv engine shouldn't error on files opened in binary.
+    parser = all_parsers
+
+    fpath = os.path.join(csv_dir_path, fname)
+
```
Comment (jreback) [inline pandas/tests/io/parser/test_common.py]: can you compare these versus the file read w/o open it?

something like this (for the 2 cases)

```
expected = parser.read_csv(fpath....)
result = .....
tm.assert_frame_equal(....)

result = ....
tm.assert....
```

**Review by jreback — CHANGES_REQUESTED** [review]

**Comment by fiendish** [comment]
> this is hard to follow, can you just use BufferedIOBase here rather than no_close?

I figured this way was more general in case there were future issues discovered beyond just BufferedIOBase, but consider it done.

**Review by TomAugspurger — APPROVED** [review]

**Comment by TomAugspurger** [comment]
Thanks @fiendish!
